### PR TITLE
ROE-1284: page 'Tell us about the trust'

### PR DESCRIPTION
**Pull request by dgolubevs**
### JIRA link
https://companieshouse.atlassian.net/browse/ROE-1284

### Change description
Added first trusts page. For this, added:
New controller
New page models
New route endpoint
New template
Fix for jestconfig

Note that the mapper style is different to that previously used in this app.

### Work checklist

- [x] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/model/Trust/page.model.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+type TrustDetails = {
```
Comment (mouhajer-ch) [inline src/model/Trust/page.model.ts:1]: Use the same file naming and structure of other model

File: src/routes/index.ts [hunk 2]
```diff
@@ -118,6 +122,20 @@ router.get(config.MANAGING_OFFICER_CORPORATE_URL + config.REMOVE + config.ID, au
 router.get(config.TRUST_INFO_URL, authentication, navigation.hasBOsOrMOs, trustInformation.get);
 router.post(config.TRUST_INFO_URL, authentication, navigation.hasBOsOrMOs, ...validator.trustInformation, checkTrustValidations, trustInformation.post);
 
+if (isActiveFeature(FEATURE_FLAG_ENABLE_TRUSTS_WEB)) {
```
Comment (mouhajer-ch) [inline src/routes/index.ts:125]: Use middleware chained for this endpoint or add it on the controller, keep this file clean

File: src/service/TrustMapperToPageService.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import { TrustHistoricalBeneficialOwner } from '@companieshouse/api-sdk-node/dist/services/overseas-entities/types';
```
Comment (mouhajer-ch) [inline src/service/TrustMapperToPageService.ts:1]: These are not services, create an util for the trust work on the utils directory and rename the file please.

Use the model defined locally, consistency please

File: src/service/TrustMapperToPageService.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import { trustType } from 'model';
+import * as Page from 'model/Trust/page.model';
+
+interface TrustMapperToPageService {
```
Comment (mouhajer-ch) [inline src/service/TrustMapperToPageService.ts:5]: Refactor this code to be an util and again keep it simple, please

File: src/service/TrustMapperToSessionService.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import { TrustHistoricalBeneficialOwner } from '@companieshouse/api-sdk-node/dist/services/overseas-entities/types';
```
Comment (mouhajer-ch) [inline src/service/TrustMapperToSessionService.ts:1]: Import types from local model

File: views/trust-details.html [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        {{ govukInsetText({
+          html: '
+            <h2 class="govuk-heading-m">Information shown on the public register</h2>
+            <p>None of the information you provide about the trust will be shown on the public Register of Overseas Entities. However, it may be shared with HMRC.</p>
+          '
+        }) }}
+
+        <input type="hidden" name="id" value="{{ pageData.id }}" />
+
+        {{ govukButton({
```
Comment (mouhajer-ch) [inline views/trust-details.html:93]: You can import it from `{% include "includes/continue-button.html" %}`

File: src/service/TrustMapperToSessionService.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import { trustType } from 'model';
+import * as Page from 'model/Trust/page.model';
+
+interface TrustMapperToSessionService {
```
Comment (mouhajer-ch) [inline src/service/TrustMapperToSessionService.ts:5]: It is not a service, could you please refactor the code

File: src/routes/index.ts [hunk 8]
```diff
@@ -30,11 +30,15 @@ import {
   signOut
 } from "../controllers";
 
+import { trustDetailsController } from '../controllers/trust.details.controller';
```
Comment (mouhajer-ch) [inline src/routes/index.ts:33]: Import from index

**Review by mouhajer-ch — COMMENTED** [review]
Could you please:

1. Make sure we have got a good description and the story link on the PR when raised.
2. Make sure that we add **unit/integration tests**, and coverage is above 90%, even if limit is 80.
3. Make sure any `node/npm` version is agreed. We could use the 16, I think, but for now we keep at node 12 (Revert please `package-lock.json` file). We can discuss/change in a second time

File: src/routes/index.ts [hunk 9]
```diff
@@ -118,6 +122,20 @@ router.get(config.MANAGING_OFFICER_CORPORATE_URL + config.REMOVE + config.ID, au
 router.get(config.TRUST_INFO_URL, authentication, navigation.hasBOsOrMOs, trustInformation.get);
 router.post(config.TRUST_INFO_URL, authentication, navigation.hasBOsOrMOs, ...validator.trustInformation, checkTrustValidations, trustInformation.post);
 
+if (isActiveFeature(FEATURE_FLAG_ENABLE_TRUSTS_WEB)) {
+  router.get(
+    config.TRUST_DETAILS_URL,
+    authentication,
```
Comment (mouhajer-ch) [inline src/routes/index.ts:128]: You need to add navigation checks, to avoid user jumping to the following page

File: src/routes/index.ts [hunk 10]
```diff
@@ -118,6 +122,20 @@ router.get(config.MANAGING_OFFICER_CORPORATE_URL + config.REMOVE + config.ID, au
...
+    trustDetailsController.get,
+  );
+
+  router.post(
+    config.TRUST_DETAILS_URL,
+    authentication,
```
Comment (mouhajer-ch) [inline src/routes/index.ts:134]: You need to add validation checks middleware if you want to catch errors

File: src/routes/index.ts (hunk 2 above)
Comment (dgolubevs) [reply src/routes/index.ts:125]: What is benefit of middleware?

File: src/routes/index.ts (hunk 8 above)
Comment (dgolubevs) [reply src/routes/index.ts:33]: What is benefit?

File: src/routes/index.ts (hunk 10 above)
Comment (dgolubevs) [reply src/routes/index.ts:134]: it is other story

File: src/controllers/trust.details.controller.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+import { trustMapperToSessionService, TrustMapperToSessionService } from '../service/TrustMapperToSessionService';
+import { Trust } from 'model/trust.model';
+
+interface TrustDetailsController {
+  get: RequestHandler;
+  post: RequestHandler;
+}
+
+const TrustDetailsControllerFactory = (
+  mapperToPageSrv: TrustMapperToPageService,
```
Comment (mwestacott) [inline src/controllers/trust.details.controller.ts:17]: Would be inclined to not shorten these variable names. Just have `mapperToPageService` and `mapperToSessionService`.

File: views/includes/inputs/checkbox-input.html [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{{ govukCheckboxes({
```
Comment (mwestacott) [inline views/includes/inputs/checkbox-input.html:1]: These look like common widget HTML definitions. Could/should they be used by all the existing forms too or are they specific to trusts?

File: src/controllers/trust.details.controller.ts [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+import * as config from '../config';
+import { logger } from '../utils/logger';
+import { getApplicationData, setExtraData } from '../utils/application.data';
+import { TrustMapperToPageService, trustMapperToPageService } from '../service/TrustMapperToPageService';
+import { ApplicationData } from '../model/application.model';
+import * as PageModel from '../model/Trust/page.model';
+import { trustMapperToSessionService, TrustMapperToSessionService } from '../service/TrustMapperToSessionService';
+import { Trust } from 'model/trust.model';
+
+interface TrustDetailsController {
```
[the remark this answers is not in the document]
Comment (SteveHicksCH) [reply src/controllers/trust.details.controller.ts:11]: > Why this boilerplate code, could you please follow the pattern and structure already defined on other controllers

Can you provide more detail on what you class as "boilerplate code" using the file src/controllers/beneficial.owner.individual.controller.ts as an example

File: src/controllers/trust.details.controller.ts [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    logger.debugRequest(req, `${req.method} ${req.route.path}`);
+
+    const trustId = req.params['id'];
+
+    //  get trust data from session
+    const appData: ApplicationData = getApplicationData(req.session);
+    const trusts: Trust[] = appData[TrustKey] ?? [];
+    const trustIdx = trusts.findIndex(trust => trust.trust_id === trustId);
+
+    if (trustIdx >= 0) {
```
Comment (markpit) [inline src/controllers/trust.details.controller.ts:64]: could you also add a comment above line 64 similar to the other steps in this function, just to say something like 
`// remove trust from the array if it is present in the trusts array`

**Comment by markpit** [comment]
I think the package-lock.json can be removed from the PR as I can no longer see any changes to package.json

File: src/utils/trust/mapper.to.page.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import * as Trust from 'model/trust.model';
+import * as Page from 'model/trust.page.model';
+
+const mapDetail = (
```
Comment (markpit) [inline src/utils/trust/mapper.to.page.ts:4]: naming - can you rename mapDetail function name to 'mapToTrustDetails' as the function will return a TrustDetails object.

Can you also do the same for mapBeneficialOwner - 'mapToTrustBeneficialOwner'

File: src/utils/trust/mapper.to.session.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import * as Trust from 'model/trust.model';
+import * as Page from 'model/trust.page.model';
+
+const mapDetail = (
```
Comment (markpit) [inline src/utils/trust/mapper.to.session.ts:4]: naming - can you rename mapDetail to be 'mapToTrust' as it will accept TrustDetails as param and return Trust

File: src/utils/trust/mapper.to.session.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+  data: Page.TrustDetails,
```
Comment (markpit) [inline src/utils/trust/mapper.to.session.ts:5]: trailing ',' on line 5 (params) and line 18 (exports) can be removed

File: src/utils/trust/mapper.to.page.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    createdDateMonth: data.creation_date_month,
+    createdDateYear: data.creation_date_year,
+    beneficialOwners: data.HISTORICAL_BO?.map((bo, idx) => mapBeneficialOwner(bo, idx)) ?? [],
+    hasAllInfo: data.unable_to_obtain_all_trust_info,
+  };
+};
+
+const mapBeneficialOwner = (
+  data: Trust.TrustHistoricalBeneficialOwner | undefined,
+  idx: number,
```
Comment (markpit) [inline src/utils/trust/mapper.to.page.ts:24]: naming - can you rename idx param to be 'index'

File: src/model/trust.page.model.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
+type TrustDetails = {
```
Comment (ch-tstrickland) [inline src/model/trust.page.model.ts:1]: This seems to be a different way of modelling this to other bits of the model, e.g.
beneficial.owner.individual.model.ts
https://github.com/companieshouse/overseas-entities-web/blob/main/src/model/beneficial.owner.individual.model.ts

Please could you make it more like that?

File: src/controllers/trust.details.controller.ts [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+  res: Response,
+  next: NextFunction,
+): void => {
+  try {
+    logger.debugRequest(req, `${req.method} ${req.route.path}`);
+
+    const appData: ApplicationData = getApplicationData(req.session);
+
+    const trustId = req.params['id'];
+    const pageData: PageModel.TrustDetails = mapperToPageService.mapDetail(
```
Comment (ch-tstrickland) [inline src/controllers/trust.details.controller.ts:26]: This appears to be using explicit mapping, as opposed to the implicit mapping used elsewhere in the app.
My gut instinct is that I prefer explicit mapping (it breaks quickly and loudly when the mapping code ought to change, as opposed to sometimes failing silently)

However, I would much rather be consistent with the rest of the app, e.g.
https://github.com/companieshouse/overseas-entities-web/blob/main/src/controllers/beneficial.owner.individual.controller.ts#L55

File: views/beneficial-owner-type.html [hunk 21]
```diff
@@ -112,12 +112,18 @@ <h2 class="govuk-heading-l govuk-!-margin-0">What <span class="govuk-visually-hi
...
           {% if (hasTrusts) %}
-            {% set nextPage = "trust-information" %}
+            {% set nextPage = pageParams.urlToTrust %}
           {% else %}
-            {% set nextPage = "check-your-answers" %}
+            {% set nextPage = pageParams.urlToCheckYourAnswers %}
           {% endif %}
 
-          <a href= "/register-an-overseas-entity/{{ nextPage }}" role="button" draggable="false" class="govuk-button govuk-button--secondary govuk-!-margin-top- 6" data-module="govuk-button" id="submit">No more <span class="govuk-visually-hidden"> {{ hiddenTextValue }} </span> to add</a>
+          <a href= "{{ nextPage }}"
```
Comment (ch-tstrickland) [inline views/beneficial-owner-type.html:120]: As best as I can understand it, this looks like an elegant improvement. However, can we please aim for consistency with the rest of the app.

**Review by ch-tstrickland — COMMENTED** [review]
I REALLY want to take on improvements to how we do things & have agreement from other people in management that we can channel that into a separate platonic ideal app with a few pages that shows an agreed set of best practice. We can point to that and:
for new projects just point to the platonic ideal app and say "do it that way"
for existing apps backport proven ways of working as tech-debt into those apps.

In short, please can we make this all effectively the same as the rest of the app, with the possible exception of trust.page.model.ts, which I will open for discussion now.

File: src/model/trust.page.model.ts (hunk 19 above)
Comment (ch-tstrickland) [reply src/model/trust.page.model.ts:1]: By which I mean, explicit separate keys, rather than implicit.

Comment (ch-tstrickland) [inline src/model/trust.page.model.ts:1]: A question, for @markpit @mouhajer-ch @mwestacott @SteveHicksCH @dgolubevs @JoshFen95 ...

The current way that the rest of the app works is to have a model for the page that is also used for the session.
In this PR, there is a slight difference in that way of working, but there may be an argument for keeping this difference, as show below. Please share your thoughts.

We need a new page object: trust.page.model.ts
We could store this in the session and then alter the node sdk to cope with the second way of modelling trusts. The SDK would either use a feature flag, passed in, or look at the session and see one or other type of trust and use whichever one is present (erroring in the event that both are present, which should be impossible).

OR...
We could take the page model and map it to the existing trust.model.ts and store THAT in the session. In this case no changes are needed in the SDK.
At the end of the project, we can either:
keep it as is, only refactoring to get rid of the page that uses trust.model.ts, but continuing to use that as our session object.
OR
refactor to: 1) get rid of the trust.model.ts 2) use trust.page.model.ts as the session object, pass that into the SDK and 3) refactor the SDK.

The trade-off is between:
1) less work now, less churn now, with a little more work later to refactor
2) more work now (starting from this PR's starting point) and more churn with each milestone, but no need to refactor at the end.

File: views/beneficial-owner-type.html (hunk 21 above)
Comment (dgolubevs) [reply views/beneficial-owner-type.html:120]: There is no functional improvement, more improve readability
Line is too long

File: src/model/trust.page.model.ts (hunk 19 above)
Comment (dgolubevs) [reply src/model/trust.page.model.ts:1]: > refactor to: 1) get rid of the trust.model.ts 2) use trust.page.model.ts as the session object, pass that into the SDK and 3) refactor the SDK.

in this case it will work like now (

My approach separates a session object from a page object.
In current implementation session object used for page object and in other way

> refactoring
it is too much work and this is not cost it

File: src/routes/index.ts [hunk 22]
```diff
@@ -118,6 +121,15 @@ router.get(config.MANAGING_OFFICER_CORPORATE_URL + config.REMOVE + config.ID, au
...
 router.post(config.TRUST_INFO_URL, authentication, navigation.hasBOsOrMOs, ...validator.trustInformation, checkTrustValidations, trustInformation.post);
 
+router
+  .route(config.TRUST_DETAILS_URL + config.ID + '?')
+  .all(
+    isFeatureEnabled(FEATURE_FLAG_ENABLE_TRUSTS_WEB),
+    authentication,
+  )
+  .get(trustDetails.get)
+  .post(trustDetails.post);
```
[the remark this answers is not in the document]
Comment (dgolubevs) [reply src/routes/index.ts:131]: Yes, I will. But on current moment i will block page.
We will implement add BO in next slice/sprint and after this will add this check middleware

File: src/utils/trust/mapper.to.session.ts (hunk 17 above)
Comment (dgolubevs) [reply src/utils/trust/mapper.to.session.ts:5]: I would prefer to keep
https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8

https://blog.logrocket.com/best-practices-using-trailing-commas-javascript/

File: src/routes/index.ts (hunk 10 above)
Comment (ch-tstrickland) [reply src/routes/index.ts:128]: Yup, what Dmitry said - This has not being forgotten and is being dealt with in a future story.

File: dev.docker-compose.yaml [hunk 23]
```diff
@@ -0,0 +1,22 @@
+version: "3.8"
```
[the remark this answers is not in the document]
Comment (ch-tstrickland) [reply dev.docker-compose.yaml]: resolved by removal.

File: src/app.ts [hunk 24]
```diff
@@ -41,7 +38,7 @@ nunjucksEnv.addGlobal("PIWIK_START_GOAL_ID", config.PIWIK_START_GOAL_ID);
 nunjucksEnv.addGlobal("MATOMO_ASSET_PATH", `//${config.CDN_HOST}`);
 
 app.use(express.json());
-app.use(express.urlencoded({ extended: false }));
+app.use(express.urlencoded({ extended: true }));
```
[the remark this answers is not in the document]
Comment (ch-tstrickland) [reply src/app.ts]: @dgolubevs I think that you may have missed this one.

**Comment by ch-tstrickland** [comment]
> Could you please:
> 
> 1. Make sure we have got a good description and the story link on the PR when raised.
> 2. Make sure that we add **unit/integration tests**, and coverage is above 90%, even if limit is 80.
> 3. Make sure any `node/npm` version is agreed. We could use the 16, I think, but for now we keep at node 12 (Revert please `package-lock.json` file). We can discuss/change in a second time
We could probably use engine config to enforce proper node & npm versioning:
https://chrisshepherd.me/posts/enforcing-nodejs-npm-yarn-versions

File: package.json [hunk 25]
```diff
@@ -30,7 +30,7 @@
     "cookie-parser": "^1.4.6",
     "express": "^4.17.3",
     "express-validator": "^6.14.0",
-    "govuk-frontend": "^4.0.1",
+    "govuk-frontend": "^4.4.0",
```
[the remark this answers is not in the document]
Comment (ch-tstrickland) [reply package.json]: Not addressed yet. Revert please, separate PR.

[the remark this answers is not in the document]
Comment (ch-tstrickland) [reply package.json]: Reverted

File: src/app.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (ch-tstrickland) [reply src/app.ts]: Ignore me, my mistake. Reverted.

File: views/beneficial-owner-type.html (hunk 21 above)
Comment (ch-tstrickland) [reply views/beneficial-owner-type.html:120]: This is just better line wrapping. OK to keep @markpit ?

File: src/model/trust.page.model.ts (hunk 19 above)
Comment (ch-tstrickland) [reply src/model/trust.page.model.ts:1]: @markpit Would you be OK to leave a brief reply to this, in line with our chat earlier?

Comment (markpit) [reply src/model/trust.page.model.ts:1]: yes ok, I think the simplest approach is to leave the API input data structure as it is, and the SDK should just act as a wrapper to the API. So that means the data in the session should be in the same structure as the API input data. It seems sensible to me to map the page data into the session/model structure that will be passed to API.

File: src/controllers/trust.details.controller.ts (hunk 20 above)
Comment (ch-tstrickland) [reply src/controllers/trust.details.controller.ts:26]: After talking with @markpit & @mwestacott, we're OK with this explicit mapping model as-is, provided that there is a comment in any explicit mappers noting that they behave differently to the mapping elsewhere in the app.

File: src/model/trust.page.model.ts (hunk 19 above)
Comment (ch-tstrickland) [reply src/model/trust.page.model.ts:1]: After talking with @markpit & @mwestacott, we're OK with this explicit mapping model as-is, provided that there is a comment in any explicit mappers noting that they behave differently to the mapping elsewhere in the app.

Comment (ch-tstrickland) [reply src/model/trust.page.model.ts:1]: After talking with @markpit , he says that he's OK with this as-is.

**Review by mwestacott — APPROVED** [review]

File: views/includes/inputs/date-input.html [hunk 26]
```diff
@@ -0,0 +1,34 @@
+{{ govukDateInput({
```
Comment (markpit) [inline views/includes/inputs/date-input.html:1]: I would move this date-input.html file to view/includes/inputs/date folder so it is in same location as the other date 'includes'

**Comment by markpit** [comment]
also, don't forgot to add a link to the Jira ticket in the PR description

**Comment by markpit** [comment]
> > also, don't forgot to add a link to the Jira ticket in the PR description
> 
> but Jira have already link to pr at the right side 'Details > development'

The Companies House PR Template has a section for the Jira ticket link, it's just the CH standard that we have to follow while working for CH.

**Comment by markpit** [comment]
The 2 sonar code smells need fixing as well

**Comment by ch-tstrickland** [comment]
> > > also, don't forgot to add a link to the Jira ticket in the PR description
> > 
> > 
> > but Jira have already link to pr at the right side 'Details > development'
> 
> The Companies House PR Template has a section for the Jira ticket link, it's just the CH standard that we have to follow while working for CH.

I could have coached Dmitry in this better, so I'll sort this now.
That said, it would have been clearer to say "please stick the link to the Jira story in the PR under the heading "JIRA link" that was put in there by the standard template"

**Review by dgolubevs — COMMENTED** [review]
Could you please recheck

File: views/includes/inputs/date-input.html (hunk 26 above)
Comment (markpit) [reply views/includes/inputs/date-input.html:1]: ok, happy to leave as it is
